---
title: "Mouse Olfactory Analysis"
author: "Zhaorong Li"
date: "Jan/13/2022"
output: 
  
  html_document:
    toc: true
    toc_float:
      toc_collapsed: true
      toc_depth: 4
    number_sections: true
    fig_caption: yes
    theme: cerulean
editor_options: 
  chunk_output_type: console
params:
  TF: ''
---


```{r setup, include=FALSE}
library(Seurat)
library(tidyr)
library(reshape2)
library(reactable)
library(RColorBrewer)
library(psych)
library(gridExtra)
library(dplyr)
```

# MERFISH Image visualization

This dataset [1] is downloaded from https://zenodo.org/record/5745947#.YeB7VtHMKUk (Set 1). The Watershed segmentation is used in the analysis.

## Transcriptome Analysis

In this case the authors do not provide cell type assignment. So we need to use the data to generate cell type.
```{r Transcriptome analysis, fig.width=10,fig.height=10}

Count.matrix=read.csv('barcodes_per_feature.csv.gz',header = T,row.names =1,check.names = F)
Count.matrix=t(as.matrix(Count.matrix))
meta.data=read.csv('feature_metadata.csv.gz',header = T,row.names = 1,check.names = F)
Count.matrix=Count.matrix[!grepl('Blank',rownames(Count.matrix)),]
Seurat_Object=CreateSeuratObject(Count.matrix,meta.data=meta.data)

VlnPlot(Seurat_Object,c('nFeature_RNA','nCount_RNA','volume'),pt.size = 0)

Seurat_Object=subset(Seurat_Object,subset=nFeature_RNA>15&nFeature_RNA<100)

Seurat_Object=NormalizeData(Seurat_Object)
Seurat_Object=FindVariableFeatures(Seurat_Object)
Seurat_Object=ScaleData(Seurat_Object)
Seurat_Object=RunPCA(Seurat_Object,npcs = 50)

Seurat_Object=RunUMAP(Seurat_Object,dims = 1:20)

Seurat_Object=FindNeighbors(Seurat_Object,dims = 1:20)

Seurat_Object=FindClusters(Seurat_Object,resolution = 0.5,algorithm = 2)

Seurat_Object.markers=FindAllMarkers(Seurat_Object,only.pos = T)
getPalette = colorRampPalette(brewer.pal(8, "Dark2"))


color=getPalette(length(unique(Seurat_Object@meta.data$seurat_clusters)))

DimPlot(Seurat_Object,cols = color,label = T,pt.size = 1)

Seurat_Object.markers %>%
    group_by(cluster) %>%
    top_n(n = 5, wt = avg_log2FC) -> top10
DoHeatmap(Seurat_Object, features = top10$gene,group.colors = color,label = F)


Cell_spacial_information=data.frame(cell=rownames(Seurat_Object@meta.data),
                                    x=Seurat_Object@meta.data$center_x,
                                    y=Seurat_Object@meta.data$center_y,
                                    Celltype=Seurat_Object@meta.data$seurat_clusters)



```


## Plot Cell by centroid Coordinates
```{r Plot Cell by centroid Coordinates, fig.width=10,fig.height=10}

ggplot(Cell_spacial_information,aes(x=x,y=y,colour=Celltype))+geom_point()+
  xlim(min(Cell_spacial_information$x),max(Cell_spacial_information$x))+
  ylim(min(Cell_spacial_information$y),max(Cell_spacial_information$y)) + scale_color_manual(values = color)+ theme_bw() + 
  theme(
    # get rid of panel grids
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    # Change plot and panel background
    plot.background=element_rect(fill = "black"),
    panel.background = element_rect(fill = 'black'),
    # Change legend 
    #legend.position = c(0.6, 0.07),
    #legend.direction = "horizontal",
    legend.background = element_rect(fill = "black", color = NA),
    legend.key = element_rect(color = "gray", fill = "black"),
    legend.title = element_text(color = "white"),
    legend.text = element_text(color = "white")
  )

```

# Analysis

## Area Calculator Function

In this case instead of calculating the area overlap between cells and fragments, the number of cells are calculated using the coordinates of the cells.


```{r Area Cellular composition Calculator}
source('functions.R')

```

```{r Area determination and Calculation}
Area.50.50=Area_Counter(min(Cell_spacial_information$x),
                        max(Cell_spacial_information$x),
                        min(Cell_spacial_information$y),
                        max(Cell_spacial_information$y),numbers = 50,cell_coordinates = Cell_spacial_information)





reactable(Area.50.50)


```

## Normalization and Scaling

The cell composition of each fragment is scaled by the total number of cells in the fragment. The scaled number is multiplied by 1,000,000 and then log transformed.

Basically it is log transformed CPM normalized. This step is included in the Area_Cluster() function.

```{r Normalization and Scaling}

Area.50.50.normalized=Area.50.50/rowSums(Area.50.50)
Area.50.50.normalized=Area.50.50.normalized*(1e6)
Area.50.50.normalized=log(Area.50.50.normalized+1)

reactable(Area.50.50.normalized)

```

## Clustering

In this case I used the Louvain Community Clustering Algorithm from igraph package to cluster the area.


```{r clustering}

clusterlouvain <- Area_Cluster(Area.50.50,ScaleFactor = 1e6,0.5)

```

## Visualization of Clustering Results

```{r Visualization of Clustering Results, fig.width=20,fig.height=20}

Area.Cluster.assignment=clusterlouvain$Area.Cluster.assignment

Clustering.Distribution=clusterlouvain$Clustering.Distribution

melt=Area.Cluster.assignment
melt$cluster=as.factor(melt$cluster)
melt=melt(melt,id.vars='cluster')
colnames(melt)=c('Cluster','CellType','Cells')
Cluster_cell_composition=ggplot(melt,aes(x=Cells,y=Cluster,fill=CellType))+geom_bar(stat = 'identity',position = 'fill')+ scale_fill_manual(values = color)+ theme_bw()

centroid.graph=ggplot(Cell_spacial_information,aes(x=x,y=y,colour=Celltype))+geom_point()+
  xlim(min(Cell_spacial_information$x),max(Cell_spacial_information$x))+
  ylim(min(Cell_spacial_information$y),max(Cell_spacial_information$y)) + scale_color_manual(values = color)+ theme_bw() +xlab('x') +ylab('y') +
  theme(legend.position = "none",
    # get rid of panel grids
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    # Change plot and panel background
    # Change legend 
    #legend.position = c(0.6, 0.07),
    #legend.direction = "horizontal",
)
Clustering.Distribution.plot=clusterlouvain$Clustering.Distribution.plot

grid.arrange(Clustering.Distribution.plot, centroid.graph,Cluster_cell_composition, ncol=2)

```

In this case I am not familiar with the gene markers for different types of Neurons in the Olfactory bulb so I can only take the analysis to this level.

Prepping myself for the further analysis now.




# Conclusion

Coming soon.

# Reference

[1] I-Hao Wang, Violeta DurÃ¡n-Laforet, Travis E. Faust, Christina E. Baer, Dorothy P. Schafer, & Paul L. Greer. (2021). Reconstruction of the Mouse Olfactory Glomerular Map using Single-Cell Technologies [Data set]. Zenodo. https://doi.org/10.5281/zenodo.5745947










